### DISCUSSION TITLE
support for  Linux Huge Pages

### DISCUSSION THREAD

**Discussion by jakubhajek** [cloudnative-pg/cloudnative-pg#1571] [General]
Hello Team!

Thank you so much for developing and maintaining that product. I really appreciate your work and contribution to it. 

I have a general question concerning the HugePages configuration. Is the Linux Huge Pages supported by the operator? If so, what are your recommendations in regard to configuring and tuning K8S worker nodes to support that feature?

**Comment by baurmatt** [comment]
Hey @jakubhajek, did you find a solution? I'm currently investigating into this direction as well. I've came up with the following steps that need to be done to have huge page support:

* Get user overridden PostgreSQL parameter values
* Get default PostgreSQL parameter values
* Calculate huge pages with `shared_memory_size_in_huge_pages`
* `sysctl -w vm.nr_hugepages=$hugePageCount`
* `systemctl restart kubelet`
* Configure `resources.limits.hugepages-2Mi` (and additionally `resources.limits.cpu` or `resources.limits.memory` :shrug:)
* Start PostgreSQL pod

I'm currently evaluating if the new CNPG-I plugins are a valid option to implement this.

**Comment by gbartolini** [comment]
Did you see this already? https://cloudnative-pg.io/documentation/current/troubleshooting/#error-while-bootstrapping-the-data-directory

Huge pages shouldn't be anything special in terms of configuration. If you think otherwise, though, feel free to propose a change in the docs with your findings.

**Comment by jakubhajek** [comment]
Yes, I managed that. I did the same as you have described, @baurmatt. 
I prepared the base AMI, since we use EKS and configured huge pages accordingly based on the node size. 

I configured `vm.nr_hugepages` using the following formula based on the shared buffer size: 

```sh
((shm size + (BLKSZ - (shm size % BLKSZ))) / (huge page size * 1024)) + 1 = count of of hpg (no unit)
```
For the shared buffers set to 20GB, the calculation is following:
```sh
 ((21965570048 + (8192 - (21965570048 % 8192))) / (2048 * 1024)) + 1 = 10475
```

Based on that we can configure `vm.nr_hugepages` to 10475, which actually reserved 10475 pages of the 2Mi size which is ~20950MB.

Moving forward; the EKS cluster consist of a few node groups based on the configuration of the huge page, so we can deploy different database workloads depending on the memory (hugepages) needs, eg. hpg-low, hpg-medium, hpg-high.
The nodes are tainted using the appropriate labels referring to the memory configuration. 
Then in the cluster spec configuration tolerations have to be added according to schedule the instance on the appropriate node, see the example:

```jsonnet 
{
 key: 'hpg-high',
 value: 'enabled',
 operator: 'Equal',
 effect: 'NoSchedule',
},
```
The resource configuration looks following: 

```jsonnet
resources: {
 limits: {
    memory: '37Gi',
    'hugepages-2Mi': '20950Mi',
     },
 requests: {
     memory: '37Gi',
     cpu: '15',
     'hugepages-2Mi': '20950Mi',
      },
},
```
That works fine for us. Additionally, we added a Grafana dashboard to see what is the huge pages memory usage using the metrics from node exporter, eg. `node_memory_HugePages_Total`, `node_memory_HugePages_Free` etc. I can share more details about that. 

It is worth noting how the resource capacity looks from the node perspective once you configure huge pages. Let's take a look at the 4xlarge EC2 instance with 64 Gb of memory. Once a node is added to the EKS cluster the available capacity is: 

```sh
Capacity:                                                                                                                                                                                                      │
 cpu:                16                                                                                                                                                                                       │                                                                                                                                                                               
 hugepages-1Gi:      0                                                                                                                                                                                        │
 hugepages-2Mi:      20950Mi                                                                                                                                                                                  │
 memory:             64776308Ki                                                                                                                                                                               │
```

Looks good so far, there is 64 GB of memory and 20GB of hpg-2Mi, which is exactly what we configured.  Let's take a look at what can be allocated by CNPG workloads:

```sh
Allocatable:                                                                                                                                                                                                   │
 cpu:                15890m                                                                                                                                                                                   │
 hugepages-1Gi:      0                                                                                                                                                                                        │
 hugepages-2Mi:      20950Mi                                                                                                                                                                                  │
 memory:             40324212Ki                                                                                                                                                                               │
```    

The available memory is 40GB, which make sense since we have allocated 20GB for hpg-2Mi.  Take into account those values configuring resources for the PG database in the cluster spec. 

Hope that helps, I am all ears to listen to how other community members deal with that. 

@gbartolini, do you think is it worth enriching the CNPG docs with that pieces of information?
